### ISSUE TITLE
NoDuplicatesBatchSampler __len__ method returns higher number of batches than the actual one

### ISSUE THREAD

**Issue by chunlexiong-ai** [huggingface/sentence-transformers#3348]
In class NoDuplicatesBatchSampler's \_\_len\_\_ method definition, we simply use the dataset sample size to calculate the number of batches.

```python
def __len__(self) -> int:
        if self.drop_last:
            return len(self.dataset) // self.batch_size
        else:
            return (len(self.dataset) + self.batch_size - 1) // self.batch_size
```

However, when a dataset is relatively dense, we can have duplicates clashing and thus we cannot use all the samples to form batches if drop_last=True. This leads to inconsistency between len(batch_sampler) and the actual number of batches. This inconsistency will propagate to SentenceTransformerTrainer. The trainer expects more batches to come, but the batches have been exhausted. Then the training process hangs there until timeout error is thrown.

My temporary solution is to create a subclass of NoDuplicatesBatchSampler and override its \_\_len\_\_ method. And then I create a subclass of SentenceTransformerTrainer to override its get_batch_sample() method. We have to use exactly the same logic to calculate the number of batches and form the batches. Please let me know whether you'd like me to create a pull request to fix this issue. I only need to modify the NoDuplicatesBatchSampler class.

```python
class CustomNoDuplicatesBatchSampler(NoDuplicatesBatchSampler):
    def __init__(
        self,
        dataset: Dataset,
        batch_size: int,
        drop_last: bool,
        valid_label_columns: list[str] = [],
        generator: torch.Generator = None,
        seed: int = 42,
    ) -> None:
        """
        Re-define the __init__ method
        """
        super().__init__(dataset, batch_size, drop_last)
        if label_columns := set(dataset.column_names) & set(valid_label_columns):
            dataset = dataset.remove_columns(label_columns)
        self.dataset = dataset
        self.batch_size = batch_size
        self.drop_last = drop_last
        self.generator = generator
        self.seed = seed
        self._num_batches = self._compute_num_batches()


    def _compute_num_batches(self) -> int:
        """
        Compute the number of batches using the same logic and random sampling as the iterator.
        """
        if self.generator and self.seed:
            self.generator.manual_seed(self.seed + self.epoch)

        # We create a dictionary to None because we need a data structure that:
        # 1. Allows for cheap removal of elements
        # 2. Preserves the order of elements, i.e. remains random
        num_batches = 0
        remaining_indices = dict.fromkeys(torch.randperm(len(self.dataset), generator=self.generator).tolist())
        while remaining_indices:
            batch_values = set()
            batch_indices = []
            for index in remaining_indices:
                sample_values = {
                    value
                    for key, value in self.dataset[index].items()
                    if not key.endswith("_prompt_length") and key != "dataset_name"
                }
                if sample_values & batch_values:
                    continue

                batch_indices.append(index)
                if len(batch_indices) == self.batch_size:
                    num_batches += 1
                    break

                batch_values.update(sample_values)

            else:
                # NOTE: some indices might still have been ignored here
                if not self.drop_last and batch_indices:
                    num_batches += 1

            for index in batch_indices:
                del remaining_indices[index]

        return num_batches

    def __len__(self) -> int:
        return self._num_batches

class CustomSentenceTransformerTrainer(SentenceTransformerTrainer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def get_batch_sampler(
        self,
        dataset: Dataset,
        batch_size: int,
        drop_last: bool,
        valid_label_columns: list[str] | None = None,
        generator: torch.Generator | None = None,
    ) -> BatchSampler | None:
        """
        Returns the appropriate batch sampler based on the ``batch_sampler`` argument in ``self.args``.
        This batch sampler class supports ``__len__`` and ``__iter__`` methods, and is used as the ``batch_sampler``
        to create the :class:`torch.utils.data.DataLoader`.

        .. note::
            Override this method to provide a custom batch sampler.

        Args:
            dataset (Dataset): The dataset to sample from.
            batch_size (int): Number of samples per batch.
            drop_last (bool): If True, drop the last incomplete batch if the dataset size
                is not divisible by the batch size.
            valid_label_columns (List[str]): List of column names to check for labels.
                The first column name from ``valid_label_columns`` found in the dataset will
                be used as the label column.
            generator (torch.Generator, optional): Optional random number generator for shuffling
                the indices.
        """
        if isinstance(dataset, IterableDataset):
            if self.args.batch_sampler != BatchSamplers.BATCH_SAMPLER:
                logger.warning("When using an IterableDataset, you cannot specify a batch sampler.")
            return None

        if self.args.batch_sampler == BatchSamplers.NO_DUPLICATES:
            return CustomNoDuplicatesBatchSampler(
                dataset=dataset,
                batch_size=batch_size,
                drop_last=drop_last,
                valid_label_columns=valid_label_columns,
                generator=generator,
            )

        if self.args.batch_sampler == BatchSamplers.GROUP_BY_LABEL:
            return GroupByLabelBatchSampler(
                dataset=dataset,
                batch_size=batch_size,
                drop_last=drop_last,
                valid_label_columns=valid_label_columns,
            )

        if self.args.batch_sampler == BatchSamplers.BATCH_SAMPLER:
            return DefaultBatchSampler(
                SubsetRandomSampler(range(len(dataset)), generator=generator),
                batch_size=batch_size,
                drop_last=drop_last,
            )
```

**Comment by chunlexiong-ai** [comment]
> > This inconsistency will propagate to SentenceTransformerTrainer. The trainer expects more batches to come, but the batches have been exhausted. Then the training process hangs there until timeout error is thrown.
> 
> I suspected that this might have been the case, but I wasn't sure. I'm a bit hesitant to include this, as I'm wary that this will be extremely slow before training and perhaps also increase people's memory usage.

I am pretty sure the inconsistency between the actual number of batches and the returned number of batches by `__len__` is the cause of training process handing and timeout error. This is because I have observed that we the calculated number of batches by `__len__` was 1060 in my training job, the process hung at 1040, and when I applied my proposed solution, the total number of batches became exactly 1040 and the training job finished successfully. I have also done experiments to return the results of `len(trainer.get_train_dataloader())` and `len([batch for batch in batch_sampler])` using the original NoDuplicatesBatchSampler, they often return different results. Please be noted that `len(trainer.get_train_dataloader())` essentially returns the results calculated by `__len__`, and `len([batch for batch in batch_sampler])` returns the number of actually formed batches.

I understand the concern about the speed. Indeed, my proposed solution on `_len__` is essentially doing the batching, which can be slow when the dataset is big

> It's not ideal in terms of learning rate scheduling etc., but perhaps a nicer solution is to try and update `transformers` such that training does not halt if a dataloader unexpectedly runs out of data before `__len__` batches have been sampled.

This seems a better solution if it can be done. But this solution may be interconnected to other parts such as learning rate scheduling etc. I will leave this to you to decide.

> Another fix is to resample the first or last batches if the BatchSampler recognizes that it hasn't yet sampled `__len__` batches but it ran out of data?

This idea is great too. We just need to modify the sampler a bit by tracking the number of formed batches along with the sampling.

> Also, I'm not sure if the problem persists if you provide `drop_last=False` to the `NoDuplicatesBatchSampler`.

I didn't try this one because I was doing DDP. Even if I set `drop_last=False`, the process would change it to True. But I did think about what would happen if we can set `drop_last=False`. If we have lots of duplicates and set `drop_last=False`, the extreme case with current definition of NoDuplicatesBatchSampler would be we will have lots of small batches. At the end, the number of batches returned by the sampler will be greater than the calculated number of batches by `__len__`.

**Comment by chunlexiong-ai** [comment]
> Another fix is to resample the first or last batches if the BatchSampler recognizes that it hasn't yet sampled `__len__` batches but it ran out of data?

The implementation of this fix is below. Please let me know if you'd like me to put through a PR. Thanks.
class NoDuplicatesBatchSampler(SetEpochMixin, BatchSampler):
    def \_\_init\_\_(
        self,
        dataset: Dataset,
        batch_size: int,
        drop_last: bool,
        valid_label_columns: list[str] = [],
        generator: torch.Generator = None,
        seed: int = 0, # I am not sure why the default is 0, which will disable the manual_seed
    ) -> None:
        """
        This sampler creates batches such that each batch contains samples where the values are unique,
        even across columns. This is useful when losses consider other samples in a batch to be in-batch
        negatives, and you want to ensure that the negatives are not duplicates of the anchor/positive sample.

        Recommended for:
            - :class:`~sentence_transformers.losses.MultipleNegativesRankingLoss`
            - :class:`~sentence_transformers.losses.CachedMultipleNegativesRankingLoss`
            - :class:`~sentence_transformers.losses.MultipleNegativesSymmetricRankingLoss`
            - :class:`~sentence_transformers.losses.CachedMultipleNegativesSymmetricRankingLoss`
            - :class:`~sentence_transformers.losses.MegaBatchMarginLoss`
            - :class:`~sentence_transformers.losses.GISTEmbedLoss`
            - :class:`~sentence_transformers.losses.CachedGISTEmbedLoss`

        Args:
            dataset (Dataset): The dataset to sample from.
            batch_size (int): Number of samples per batch.
            drop_last (bool): If True, drop the last incomplete batch if the dataset size
                is not divisible by the batch size.
            valid_label_columns (List[str]): List of column names to check for labels.
                The first column name from ``valid_label_columns`` found in the dataset will
                be used as the label column.
            generator (torch.Generator, optional): Optional random number generator for shuffling
                the indices.
            seed (int, optional): Seed for the random number generator to ensure reproducibility.
        """
        super().__init__(dataset, batch_size, drop_last)
        if label_columns := set(dataset.column_names) & set(valid_label_columns):
            dataset = dataset.remove_columns(label_columns)
        self.dataset = dataset
        self.batch_size = batch_size
        self.drop_last = drop_last
        self.generator = generator
        self.seed = seed

    def __iter__(self) -> Iterator[list[int]]:
        """
        Iterate over the remaining non-yielded indices. For each index, check if the sample values are already in the
        batch. If not, add the sample values to the batch keep going until the batch is full. If the batch is full, yield
        the batch indices and continue with the next batch.
        """
        if self.generator and self.seed:
            self.generator.manual_seed(self.seed + self.epoch)

        # We create a dictionary to None because we need a data structure that:
        # 1. Allows for cheap removal of elements
        # 2. Preserves the order of elements, i.e. remains random
        num_batches = 0
       while num_batches < self.__len__():
            remaining_indices = dict.fromkeys(torch.randperm(len(self.dataset), generator=self.generator).tolist())
            while remaining_indices:
                if num_batches >= self.__len__():
                      break
                batch_values = set()
                batch_indices = []
                for index in remaining_indices:
                    if num_batches >= self.__len__():
                         break
                    sample_values = {
                        value
                        for key, value in self.dataset[index].items()
                        if not key.endswith("_prompt_length") and key != "dataset_name"
                    }
                   if sample_values & batch_values:
                        continue

                   batch_indices.append(index)
                   if len(batch_indices) == self.batch_size:
                        num_batches += 1
                        yield batch_indices
                        break

                    batch_values.update(sample_values)

                else:
                     # NOTE: some indices might still have been ignored here
                    if not self.drop_last and num_batches < self.__len__():
                          num_batches += 1
                          yield batch_indices

                  for index in batch_indices:
                  del remaining_indices[index]

    def __len__(self) -> int:
        if self.drop_last:
            return len(self.dataset) // self.batch_size
        else:
            return (len(self.dataset) + self.batch_size - 1) // self.batch_size

**Comment by tomaarsen** [comment]
Apologies for the delay. I'm revisiting this now with a sample script. I'm noticing that my training script does stop despite it not sampling `__len__` batches:

<details><summary>See training script</summary>

```python
from datasets import concatenate_datasets, load_dataset
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
    SentenceTransformerModelCardData,
)
from sentence_transformers.losses import MultipleNegativesRankingLoss
from sentence_transformers.training_args import BatchSamplers
from sentence_transformers.evaluation import TripletEvaluator

# 1. Load a model to finetune with 2. (Optional) model card data
model = SentenceTransformer(
    "microsoft/mpnet-base",
    model_card_data=SentenceTransformerModelCardData(
        language="en",
        license="apache-2.0",
        model_name="MPNet base trained on AllNLI triplets",
    )
)

# 3. Load a dataset to finetune on
dataset = load_dataset("sentence-transformers/all-nli", "triplet")
train_dataset = dataset["train"].select(range(20))
train_dataset = concatenate_datasets([train_dataset] * 50)

# 4. Define a loss function
loss = MultipleNegativesRankingLoss(model)

# 5. (Optional) Specify training arguments
args = SentenceTransformerTrainingArguments(
    # Required parameter:
    output_dir="models/mpnet-base-all-nli-triplet",
    # Optional training parameters:
    num_train_epochs=1,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    learning_rate=2e-5,
    warmup_ratio=0.1,
    fp16=True,  # Set to False if you get an error that your GPU can't run on FP16
    bf16=False,  # Set to True if you have a GPU that supports BF16
    batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicate samples in a batch
    dataloader_drop_last=True,
    # Optional tracking/debugging parameters:
    logging_steps=10,
    run_name="mpnet-base-all-nli-triplet",  # Will be used in W&B if `wandb` is installed
)

# 7. Create a trainer & train
trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    loss=loss,
)
trainer.train()

print("Training complete")

model.save_pretrained("models/mpnet-base-all-nli-triplet-3348")
```

</details>

Which outputs:
```
{'loss': 3.1971, 'grad_norm': inf, 'learning_rate': 2e-05, 'epoch': 0.16}
{'loss': 1.703, 'grad_norm': 16.420421600341797, 'learning_rate': 1.7090909090909092e-05, 'epoch': 0.32}                                                         
{'loss': 0.3426, 'grad_norm': 14.923328399658203, 'learning_rate': 1.3454545454545455e-05, 'epoch': 0.48}                                                        
{'loss': 0.1575, 'grad_norm': 5.990853786468506, 'learning_rate': 9.81818181818182e-06, 'epoch': 0.65}                                                           
{'loss': 0.0329, 'grad_norm': 0.8214401006698608, 'learning_rate': 6.181818181818182e-06, 'epoch': 0.81}                                                         
{'train_runtime': 13.3935, 'train_samples_per_second': 74.663, 'train_steps_per_second': 4.629, 'train_loss': 1.007838373106939, 'epoch': 0.87}                  
 87%|███████████████████████████████████████████████████████████████████████████████████████████████████████████▏               | 54/62 [00:11<00:01,  4.55it/s] 
Training complete
```

I'm on these versions:
* `transformers`: 4.51.3
* `accelerate`: 1.5.1

What versions are you on? And are you using e.g. distributed training?

- Tom Aarsen

**Comment by chunlexiong-ai** [comment]
Thanks for your response and apologies for my late reply because I was on vacation. 

I'm on these versions:

- `transformers`: 4.51.3
- `accelerate`: 1.5.2
- `sentence-transformers`: 3.4.1

Yes I am using distributed training. The timeout error was thrown via NCCL (screenshot is attached below):

![Image](https://github.com/user-attachments/assets/740ddca5-d026-4ccc-8fdf-a21558944cfd)

I didn't test it without distributed training.

**Comment by tomaarsen** [comment]
Hmm, I suspect that's a quirk by `accelerate` when using DDP. Let me try and raise this internally to discuss the best fix.

- Tom aarsen

**Comment by chunlexiong-ai** [comment]
I did a similar test without distributed training and there was no problem. Most likely the problem comes from somewhere using DDP.

**Comment by tomaarsen** [comment]
My colleague from `accelerate` has picked this up, let's hope he can resolve it directly in `accelerate`.

- Tom Aarsen

**Comment by chunlexiong-ai** [comment]
Even if he can resolve it directly in `accelerate` (let's say the training finishes as soon as the batches are exhausted, no matter what the `__len__` method returns), do you think it is a good idea to leave NoDuplicatesBatchSampler's `__len__` method returns a different number from the actual number of batches?

**Comment by tomaarsen** [comment]
I do think that a close estimate is often preferable over the true length that requires exhausting the entire training dataset and determining the entire list of batch ids prior to training. I recognize that it also affects e.g. learning rate scheduling, but in most situations, it should only be a difference of 0 to 2 steps, and only if you use `drop_last=True`.

- Tom Aarsen

**Comment by chunlexiong-ai** [comment]
Yes, you are right the difference will only be 1 or 2, depending on how dense the duplicated samples are. Looking forward to the upgrade on `accelerate` to fix this. Thanks a lot.

**Comment by chunlexiong-ai** [comment]
@tomaarsen Is there any update on the fix? 

Recently I have started working with really big dataset and lots of GPUs, which means I need to form relatively big batches. For example, I have 95744 samples in a chunk of data, and my batch size is 512. In theory I should have 95744 / 512 = 187 batches, but with NoDuplicatesBatchSampler, I only got 164 batches. The difference is 23! At the moment I am using my own patch to fix bug (first approach mentioned above). As the dataset is big, it is very slow as well. I will try the second approach before you guys fix it.

**Comment by tomaarsen** [comment]
I'm afraid I haven't heard back yet :( 
Perhaps a bit of a quicker route is opening an issue on `accelerate` directly.

- Tom Aarsen

**Comment by alireza-aminzadeh** [comment]
I'd like to work on this.

`__len__` will stay a cheap upper-bound estimate â€” precomputing the exact count before training is too expensive on large datasets, as discussed. Instead, `__iter__` will reshuffle and continue sampling when a pass yields fewer batches than `__len__` (the resample approach suggested above). That keeps DDP/accelerate in sync without a full pre-pass.

I'll open a PR with unit tests shortly.

**Comment by alireza-aminzadeh** [comment]
Opened a PR with the resample-to-length fix and unit tests: https://github.com/huggingface/sentence-transformers/pull/3946
